legacy/c3_trajectory_generator/src/node.cpp

```diff
@@ -167,22 +181,59 @@ struct Node {
     current_waypoint_t = now;
   }
 
-  void timer_callback(const ros::TimerEvent &) {
-    if (!c3trajectory) return;
+  void timer_callback(const ros::TimerEvent &)
+  {
+    if (!c3trajectory)
+      return;
 
     ros::Time now = ros::Time::now();
 
-    if (actionserver.isNewGoalAvailable()) {
+    auto old_waypoint = current_waypoint;
+
+    if (actionserver.isNewGoalAvailable())
+    {
       boost::shared_ptr<const mil_msgs::MoveToGoal> goal = actionserver.acceptNewGoal();
       current_waypoint = subjugator::C3Trajectory::Waypoint(
-          Point_from_PoseTwist(goal->posetwist.pose, goal->posetwist.twist), goal->speed,
-          !goal->uncoordinated);
+          Point_from_PoseTwist(goal->posetwist.pose, goal->posetwist.twist), goal->speed, !goal->uncoordinated, !goal->blind);
       current_waypoint_t = now;
       this->linear_tolerance = goal->linear_tolerance;
       this->angular_tolerance = goal->angular_tolerance;
+
+      // Check if waypoint is valid
+      std::pair<bool, WAYPOINT_ERROR_TYPE> checkWPResult =
+          waypoint_validity_.is_waypoint_valid(Pose_from_Waypoint(current_waypoint), current_waypoint.do_waypoint_validation);
+      if (checkWPResult.first == false)  // got a point that we should not move to
+      {
+        if (checkWPResult.second ==
+            WAYPOINT_ERROR_TYPE::UNKNOWN)  // if unknown, check if there's a huge displacement with the new waypoint
+        {
+          auto a_point = Pose_from_Waypoint(current_waypoint);
+          auto b_point = Pose_from_Waypoint(old_waypoint);
+          // If moved more than .5m, then don't allow
+          if (abs(a_point.position.x - b_point.position.x) > .5 || abs(a_point.position.y - b_point.position.y) > .5)
+          {
+            ROS_ERROR("can't move there! - need to rotate");
+            current_waypoint = old_waypoint;
+          }
+        }
+        // if point is occupied, reject move
+        if (checkWPResult.second == WAYPOINT_ERROR_TYPE::OCCUPIED)
+        {
+          ROS_ERROR("can't move there! - waypoint is occupied");
+          current_waypoint = old_waypoint;
+        }
+        // if point is above water, reject move
+        if (checkWPResult.second == WAYPOINT_ERROR_TYPE::ABOVE_WATER)
+        {
+          ROS_ERROR("can't move there! - waypoint is above water");
+          current_waypoint = old_waypoint;
+        }
+      }
     }
-    if (actionserver.isPreemptRequested()) {
+    if (actionserver.isPreemptRequested())
+    {
       current_waypoint = c3trajectory->getCurrentPoint();
+      current_waypoint.do_waypoint_validation = c3trajectory->do_waypoint_validation;
```

Comment:
You should probably always set the current waypoint to never have waypoint validation if a movement was canceled.